### DISCUSSION TITLE
Getting started

### DISCUSSION THREAD

**Discussion by lukepuplett** [panva/oauth4webapi#41] [Q&A]
Hey, it seems like a lot of hard work has gone into this and it could be what I need.

I'm building my first Deno + TypeScript + Svelte web app (I'm C# usually). I need to build a login with Google feature. In C# we have (overly large and opaque) frameworks that do this for us.

Though when OAuth was new, I did read the spec and implement it by hand in C#. It was nice because I could keep the code focused and step through it all in a debugger.

I'm looking to do the same in Deno, create the redirect URLs, handle callbacks and sign/verify parameters, exchange tokens etc. and I feel I just need a handful of helpers.

- Is this the package I need?
- If so, how do I help myself using what's available today?

I've looked through the names and documentation for the public functions and some look like they might be what I need but right now it's too much of a leap to putting it together, I need an on-ramp.

Thank you.

**Comment by panva** [comment]
Hi @lukepuplett,

It seems this is the package you need. Although I need to point out it's not really intended for direct consumption. README reads:

> This software is a collection of routines upon which framework-specific client modules may be written.

If you're well familiar with the specifications and profiles this software targets you certainly could use it directly.

Do you possibly have any concrete questions? I would be happy to answer those.

**Comment by lukepuplett** [comment]
Thank you. I'm going through each public type, it's like an OAuth Lego set, and looking at its doc and I think I can put something together. OAuth is quite a simple dance and it's lovely to fully understand what's happening and to own that code. Thanks for getting back to me!

Keep this open for a week as it's all very new. I keep losing days down other rabbit holes.

**Reply by panva** [reply]
I believe you will find the relevant bits in https://github.com/panva/oauth4webapi/blob/d6248607a09c85fce81223ff549b03244a4fa2db/examples/code.ts, just sprinkle them in your respective controllers/handlers,...

**Reply by lukepuplett** [reply]
Hi Filip,

When logging on with Google, I don't need to send a `code_verifier` since this is a web server flow where my secrets aren't exposed (no need for PKCE).

Sadly, Google won't ignore the parameter and it's not optional:

```TypeScript
export async function authorizationCodeGrantRequest(
  as: AuthorizationServer,
  client: Client,
  callbackParameters: CallbackParameters,
  redirectUri: string,
  codeVerifier: string,
  options?: TokenEndpointRequestOptions,
): Promise<Response> { ... }
```

Google's response:

```JSON
Google access token request returned: 400: Bad Request/r/n {
  "error": "invalid_grant",
  "error_description": "code_verifier or verifier is not needed."
}
```

**Reply by panva** [reply]
Are you sending a code challenge and code challenge method with your authorization request?

**Reply by lukepuplett** [reply]
Thanks for such a fast reply! No, I'm following Google's docs and they either don't support it or fail to mention the optional `code_challenge` parameter on their authorization URL template.

https://developers.google.com/identity/protocols/oauth2/web-server#creatingclient

However, it is mentioned here:

https://developers.google.com/identity/protocols/oauth2/native-app#step1-code-verifier

Presumably because it's needed for installed apps.

**Reply by panva** [reply]
Well, try and use it as shown in the code example, a unique verifier and challenge for every authorization. Using S256 PKCE is THE way to protect from xsrf and various other attacks which is why the verifier use is made required in the grant request functions.

**Reply by lukepuplett** [reply]
I hear you! I'll send a challenge and see what happens... I would suggest making the verifier optional for server-side flows as I can imagine a bunch of implementations not supporting PKCE (and not-ignoring the additional params).

Be right back! I've got to do the school run.

**Reply by panva** [reply]
Thanks for the suggestion but the way this is done reflects the oauth profiles it targets. OAuth 2.1, OAuth 2.0 with the security bcp adhered, as well as FAPI 2.0 all mandate the use of PKCE regardless of whether it's a public client or a client with secured credentials.


I am also not claiming this module can be used with any AS, the interoperability level is given by the declared profiles and exhibiting their behaviour (conformance).

**Reply by lukepuplett** [reply]
I understand. It seems to be working (I have progressed to a new error) with PKCE codes provided, so perhaps it's Google documentation team I should get in touch with.

**Reply by lukepuplett** [reply]
I have an access token! I was stuck for a while because the help for one of the arguments for `authorizationCodeGrantRequest` is incorrect.

```TypeScript
 * @param redirectUri `redirect_uri` value used in the authorization request.
 ```

It actually wanted the URL to callback my app with that's registered with Google. Thanks.

**Reply by panva** [reply]
But that is actually correct, isn't it? Isn't the value exactly the one you pre-registered and therefore used as the `redirect_uri` parameter at the authorization server's  authorization endpoint? (e.g. in the authorization request?).

**Reply by panva** [reply]
> so perhaps it's Google documentation team I should get in touch with.

I don't think so, you just happened to use PKCE incorrectly but they exhibited the right behaviour. You didn't provide the parameters in authorization endpoint, and then probably provided a dummy value using the module's method. At their token endpoint they correctly threw `invalid_grant` because of the mismatch.

Their support for PKCE is also correctly signalled in their [discovery document](https://accounts.google.com/.well-known/openid-configuration).

**Reply by lukepuplett** [reply]
> But that is actually correct, isn't it? Isn't the value exactly the one you pre-registered and therefore used as the `redirect_uri` parameter at the authorization server's authorization endpoint? (e.g. in the authorization request?).

Oh I'm sorry. I find all the naming of all the requests and URLs confusing; there's the redirect my app makes to send the user's browser to the authorization page, and then Google also redirects back to my app with the auth code etc. Two "redirect URLs". So I thought it was the authorization URL that I redirected the user to!

On the point about Google's documentation, they don't mention PKCE at all, nor any of its query parameters, which we've discovered it actually does support.

Anyone reading Google's doc and using your library will be confused because PKCE is not mentioned at all, but is mandatory on your library. Hope that makes sense. Thanks Filip.

**Reply by panva** [reply]
Sure, notice that my sample also checks discovery result for PKCE support. https://github.com/panva/oauth4webapi/blob/d6248607a09c85fce81223ff549b03244a4fa2db/examples/code.ts#L17-L21

**Reply by lukepuplett** [reply]
Oh wow. I never saw there were samples :-D thanks!!

**Reply by panva** [reply]
Well, they're not executable samples but they show you what to do when.

**Reply by lukepuplett** [reply]
Just dropping you a note to say that I have it all working and thanks for your time today.
